File: pa5/Dictionary.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>
#include "Dictionary.h"

const int tableSize = 101;

// private types --------------------------------------------------------------

// NodeObj
typedef struct NodeObj{
    char* key;
    char* value;
    struct NodeObj* next;
} NodeObj;

// Node
typedef NodeObj* Node;

// newNode()
// constructor of the Node type
Node newNode(char* k, char* v) {
    Node N = malloc(sizeof(NodeObj));
    assert(N!=NULL);
    N->key = calloc(strlen(k)+1, sizeof(char));
    N->value = calloc(strlen(k)+1, sizeof(char));
    N->key = k;
    N->value = v;
    N->next = NULL;
    return(N);
}

// freeNode()
// destructor for the Node type
void freeNode(Node* pN){
    if( pN!=NULL && *pN!=NULL ){
        free((*pN)->key);
        free((*pN)->value);
        free(*pN);
        *pN = NULL;
    }
}

// DictionaryObj
typedef struct DictionaryObj{
    Node root;
    Node table;
    int numPairs;
} DictionaryObj;

typedef DictionaryObj* Dictionary;

// public functions -----------------------------------------------------------

// newDictionary()
// constructor for the Dictionary type
Dictionary newDictionary(void){
    Dictionary D = malloc(sizeof(DictionaryObj));
    assert(D!=NULL);
    D->table = calloc(tableSize, sizeof(DictionaryObj));
    D->root = NULL;
    D->table = NULL
    D->numPairs = 0;
    return D;
}

// freeDictionary()
// destructor for the Dictionary type
void freeDictionary(Dictionary* pD){
    if( pD!=NULL && *pD!=NULL ){
        if( !isEmpty(*pD) ) makeEmpty(*pD);
        //free(*pD->table);
        free(*pD);
        *pD = NULL;
    }
}

// rotate_left()
// rotate the bits in an unsigned int
unsigned int rotate_left(unsigned int value, int shift) {
    int sizeInBits = 8 * sizeof(unsigned int);
    shift = shift & (sizeInBits - 1);  // remainder of shift on division by sizeInBits
    if ( shift == 0 )
    return value;
    else
    return (value << shift) | (value >> (sizeInBits - shift)); // perform rotation
}

// pre_hash()
// turn a string into an unsigned int
unsigned int pre_hash(char* input) {  // input points to first char in string
    unsigned int result = 0xBAE86554;  // = 10111010111010000110010101010100
    while (*input) {                   // while *input is not '\0' (not end of string)
        result ^= *input++;                // result = result ^ *input (current char alters result)
        // input++  (go to next char)
        result = rotate_left(result, 5);   // rotate result by fixed amount
    }
    return result;  // result is now a random looking bit pattern depending on input string
}

// hash()
// turns a string into an int in the range 0 to tableSize-1
int hash(char* key){
    return pre_hash(key)%tableSize;
}
// isEmpty()
// returns 1 (true) if S is empty, 0 (false) otherwise
// pre: none
int isEmpty(Dictionary D){
    //if the Dictionary is NULL throw the error
    if( D==NULL ){
        fprintf(stderr,
                "Dictionary Error: calling isEmpty() on NULL Dictionary reference\n");
        exit(EXIT_FAILURE);
    }
    //if the number of items in the Dictionary is zero return true else return
    //false
    if(D->numPairs==0){
        return 1;
    }
    return 0;
}


// find()
// returns a reference to the Node at position index in this Hash Dictionary
Node find(Dictionary D, char* k){
    //while N does not equal null continue through the dictionary until the wanted key is found
    for(Node N = D->table[hash(k)]; N != NULL; N = N->next){
        //once the key has been found return it
        if (strcmp(N->key, k) == 0){
            return N;
        }
    }
    return NULL;
}

// size()
// returns the number of (key, value) pairs in D
// pre: none
int size(Dictionary D){
    if( D == NULL){
        fprintf(stderr, "Dictionary Error: calling size() on NULL Dictionary reference\n");
        exit(EXIT_FAILURE);
    }
    return D->numPairs;
}

// lookup()
// returns the value v such that (k, v) is in D, or returns NULL if no
// such value v exists.
// pre: none
char* lookup(Dictionary D, char* k){
    int counter = 0;
    //if the Dictionary is NULL throw an error
    if( D==NULL ){
        fprintf(stderr, "Dictionary Error: calling lookup() on NULL Dictionary reference\n");
        exit(EXIT_FAILURE);
    }
    //initialize N the key to be found so the corresponding value can be found
    Node N = find(D, k);
    if (N == NULL){
        return NULL;
    }else{
        return N->value;
    }
}

// insert()
// inserts new (key,value) pair into D
// pre: lookup(D, k)==NULL
void insert(Dictionary D, char* k, char* v){
    int counter = 0;
    int nothing = 0;
    Node N;
    if(D == NULL){
        fprintf(stderr, "Dictionary Error: cannot call insert on a NULL Dictionary");
        exit(EXIT_FAILURE);
    }
    if (lookup(D, k) != NULL){
        
        fprintf(stderr, "Dictionary Error: cannot insert duplicate keys\n");
        exit(EXIT_FAILURE);
    }
    //intialize a new Node object if current node is empty
    N = newNode(k, v);
    if(D->table[hash(k)] == NULL){
        D->table[hash(k)] = newNode(k, v);
        D->numPairs++;
    } else {
        N = newNode(k, v);
        N->next = D->table[hash(k)];
        D->table[hash(k)] = N;
        D->numPairs++;
    }
}

// delete()
// deletes pair with the key k
// pre: lookup(D, k)!=NULL
void delete(Dictionary D, char* k){
    //if the key is not found in the Dictionary then exit
    if (lookup(D,k)==NULL){
        fprintf(stderr, "Dictionary error: cannot delete non-existent key\n");
        exit(EXIT_FAILURE);
    }
    //if the current key in the dictionary does not match the key we are looking
    //for, move on.
    D->table[hash(k)] = NULL;
    D->numPairs--;
}

// makeEmpty()
// re-sets D to the empty state.
// pre: none
void makeEmpty(Dictionary D){
    //setting top/bot to NULL and items to zero empties the Dictionary
    int counter = 0;
    while(D->table[counter] != NULL && counter < tableSize){
        Node N;
        N = D->table[counter];
        D->table[counter]=N->next;
        freeNode(&N);
        N = NULL;
        counter++;
    }
    D->numPairs = 0;
    
}

void printDictionary(FILE* out, Dictionary D){
    Node N;
    if(D==NULL){
        fprintf(stderr, "Dictionary Error: printDictionary() called on NULL Dictionary reference\n");
        exit(EXIT_FAILURE);
    }
    for(int i =0; i<tableSize; i++){
        N = D->table[i];
        while(N != NULL){
            fprintf(out, "%s %s \n" , N->key, N->value);
            N = N->next;
        }
    }
}

File: Lab7/Dictionary.java
//Bora Mutluoglu | bmutluog | 1564633 | cs12M | March 5th, 2018 |
//Implementing a dictionary interface using BST, translated the c version
//into a Dictionary.java version
public class Dictionary implements DictionaryInterface {

	// private classes functions and variables
	private class Node {
		String key;
		String value;
		Node left;
		Node right;

		Node(String x, String y) {
			key = x;
			value = y;
			left = null;
			right = null;
		}
	}

	Node root;
	int numPairs;

	// findKey()
	// returns the Node containing key k in the subtree rooted at R, or returns
	// NULL if no such Node exists
	private Node findKey(Node N, String k) {
		if (N == null || k.equals(N.key)) {
			return N;
		} else if (k.compareTo(N.key) < 0) {
			return findKey(N.left, k);
		} else {
			return findKey(N.right, k);
		}
	}

	// findParent()
	// returns the parent of N in the subtree rooted at R, or returns NULL
	// if N is equal to R. (pre: R != NULL)
	Node findParent(Node N, Node R) {
		Node P = null;
		if (N != R) {
			P = R;
			while (P.left != N && P.right != N) {
				if (N.key.compareTo(P.key) < 0) {
					P = P.left;
				} else {
					P = P.right;
				}
			}
		}
		return P;
	}

	// findLeftmost()
	// returns the leftmost Node in the subtree rooted at R, or NULL if R is
	// NULL.
	Node findLeftmost(Node R) {
		Node L = R;
		if (L != null)
			for (; L.left != null; L = L.left)
				;
		return L;
	}

	// printInOrder()
	// prints the (key, value) pairs belonging to the subtree rooted at R in
	// order
	// of increasing keys to file pointed to by out.
	void printInOrder(Node R) {
		String out = "";
		if (R != null) {
			printInOrder(R.left);
			System.out.println(R.key + " " + R.value);
			printInOrder(R.right);
		}
	}

	// deleteAll()
	// deletes all Nodes in the subtree rooted at N.
	void deleteAll(Node N) {
		if (N != null) {
			root = null;
			numPairs = 0;
		}
	}

	// public functions from the Dictionary Interface mostly

	// constructor for the Dictionary class
	public Dictionary() {
		root = null;
		numPairs = 0;
	}
	// isEmpty()
	// returns 1 (true) if D is empty, 0 (false) otherwise
	// pre: none
	public boolean isEmpty() {
		if (numPairs == 0) {
			return true;
		} else {
			return false;
		}
	}
	// size()
	// returns the number of (key, value) pairs in D
	// pre: none
	public int size() {
		return numPairs;
	}
	// lookup()
	// returns the value v such that (k, v) is in D, or returns NULL if no 
	// such value v exists.
	// pre: none
	public String lookup(String key) {
		Node N;
		N = findKey(root, key);
		return (N == null ? null : N.value);
	}
	// insert()
	// inserts new (key,value) pair into D
	// pre: lookup(D, k)==NULL
	public void insert(String key, String value) throws DuplicateKeyException {
		if (findKey(root, key) != null) {
			throw new DuplicateKeyException("Dictionary Error: cannot insert duplicate keys");
		}
		Node N, A, B;
		N = new Node(key, value);
		B = null;
		A = root;
		while (A != null) {
			B = A;
			if (key.compareTo(A.key) < 0) {
				A = A.left;
			} else {
				A = A.right;
			}
		}
		if (B == null) {
			root = N;
		} else if (key.compareTo(B.key) < 0) {
			B.left = N;
		} else {
			B.right = N;
		}
		numPairs++;
	}
	// delete()
	// deletes pair with the key k
	// pre: lookup(D, k)!=NULL
	public void delete(String key) throws KeyNotFoundException {
		Node N, P, S;
		N = findKey(root, key);
		if (N == null) {
			throw new KeyNotFoundException("Dictionary Error: calling delete() on NULL Dictionary reference\n");
		}
		if (N.left == null && N.right == null) {
			if (N == root) {
				root = null;
			} else {
				P = findParent(N, root);
				if (P.right == N)
					P.right = null;
				else
					P.left = null;
			}
		} else if (N.right == null) {
			if (N == root) {
				root = N.left;
			} else {
				P = findParent(N, root);
				if (P.right == N)
					P.right = N.left;
				else
					P.left = N.left;
			}
		} else if (N.left == null) {
			if (N == root) {
				root = N.right;
			} else {
				P = findParent(N, root);
				if (P.right == N)
					P.right = N.right;
				else
					P.left = N.right;
			}
		} else {
			S = findLeftmost(N.right);
			N.key = S.key;
			N.value = S.value;
			P = findParent(S, N);
			if (P.right == S)
				P.right = S.right;
			else
				P.left = S.right;
		}
		numPairs--;
	}
	// makeEmpty()
	// re-sets D to the empty state.
	// pre: none
	public void makeEmpty() {
		deleteAll(root);
		root = null;
		numPairs = 0;
	}

	// toString()
	// pre: none
	// prints a text representation of root
	public String toString() {
		printInOrder(root);
		return "";
	}

}
